### PR TITLE
#156: add FirstLandlordSelectionPolicy

### PR DESCRIPTION
**Pull request by cai-jack**
see the two screen shots

pic1: test bid first, the other two bid later, with diamond 2 winning the bid, only test can pick up
pic2: after pickup, test is the landlord, and trump is diamond 2

### CODE REVIEW

**Comment by cai-jack** [comment]
Screen Shot 2020-06-24 at 6 33 17 PM
Screen Shot 2020-06-24 at 6 33 33 PM

File: core/src/game_state.rs [hunk 1]
```diff
@@ -1480,7 +1504,20 @@ impl DrawPhase {
             bail!("nobody has bid yet")
         } else {
             let winning_bid = bail_unwrap!(self.autobid.or_else(|| self.bids.last().copied()));
-            let landlord = self.propagated.landlord.unwrap_or(winning_bid.id);
+            let first_bid = bail_unwrap!(self.autobid.or_else(|| self.bids.first().copied()));
+            let landlord;
+
+            if self.propagated.first_landlord_selection_policy
```
Comment (rbtying) [inline core/src/game_state.rs:1510]: Perhaps,

```
landlord = match self.propagated.first_landlord_selection_policy {
  FirstLandlordSelectionPolicy::ByWinningBid => self.propagated.landlord.unwrap_or(winning_bid.id),
  FirstLandlordSelectioNPolicy::ByFirstBid => self.propagated.landlord.unwrap_or(first_bid.id),
};
```

?

Comment (cai-jack) [reply core/src/game_state.rs:1510]: let landlord = match self.propagated.first_landlord_selection_policy {
                FirstLandlordSelectionPolicy::ByWinningBid => {
                    self.propagated.landlord.unwrap_or(winning_bid.id)
                }
                FirstLandlordSelectionPolicy::ByFirstBid => {
                    if self.propagated.landlord.is_none() {
                        self.propagated.landlord.unwrap_or(first_bid.id)
                    } else {
                        self.propagated.landlord.unwrap_or(winning_bid.id)
                    }
                }
            };

File: core/src/game_state.rs [hunk 2]
```diff
@@ -1480,7 +1504,20 @@ impl DrawPhase {
...
         } else {
             let winning_bid = bail_unwrap!(self.autobid.or_else(|| self.bids.last().copied()));
-            let landlord = self.propagated.landlord.unwrap_or(winning_bid.id);
+            let first_bid = bail_unwrap!(self.autobid.or_else(|| self.bids.first().copied()));
+            let landlord = match self.propagated.first_landlord_selection_policy {
+                FirstLandlordSelectionPolicy::ByWinningBid => {
+                    self.propagated.landlord.unwrap_or(winning_bid.id)
+                }
+                FirstLandlordSelectionPolicy::ByFirstBid => {
+                    if self.propagated.landlord.is_none() {
```
Comment (rbtying) [inline core/src/game_state.rs:1513]: I don’t think you need this if statement— unwrap_or only uses the right hand side when the left hand side is None

**Comment by cai-jack** [comment]
btw, how did you manage to remove desc.d.ts and still make it work? @rbtying  Thanks.

**Comment by rbtying** [comment]
Heh, I didn’t really — something in the package configuration is messed up so that the d.ts file in node modules isn’t getting picked up. The current hack I have is just to copy the file in to the source list, so that at least there are types

**Comment by cai-jack** [comment]
yeah, just saw your changes. Thanks.

**Pull request merged**